### ISSUE TITLE
Using Docker driver in Kitchen (Windows Environment)

### ISSUE THREAD

**Issue by dshah001** [test-kitchen/kitchen-docker#113]
1. I have built docker Client on Windows as per https://ahmetalpbalkan.com/blog/compiling-docker-cli-on-windows/. 
2. I have boot2Docker where my Docker Daemon resides.
3. I have installed test-kitchen and kitchen-docker driver in windows.

After above setup, I am able to manage boot2docker images in windows environment.

C:\Users\dshah001\kitchen\docker>docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
hello-world         latest              e45a5af57b00        11 weeks ago        910 B

Issue: When I try running "kitchen list", it still cannot recognize docker CLI.

> > > > > > ------Exception-------
> > > > > > Class: Kitchen::UserError
> > > > > > ## Message: You must first install the Docker CLI tool http://www.docker.io/gettingstarted/
> > > > > > 
> > > > > > Please see .kitchen/logs/kitchen.log for more details
> > > > > > Also try running `kitchen diagnose --all` for configuration

My .kitchen.yml file is below:

---

driver:
  name: docker

provisioner:
  name: chef_solo

platforms:
- name: ubuntu-12.04
- name: centos-6.4

suites:
- name: default
  run_list:
  attributes:

Is there something I need to define in my .kitchen.yml so it can recognize my Docker CLI in Windows.

**Comment by peterabbott** [comment]
The current driver does not work on windows. It uses unix specific functionality. I had been working (with somebody else on windows) on a branch to get working on Windows. https://github.com/peterabbott/kitchen-docker/tree/windows-integration

To try it out, you can update your Gemfile as follows. Not yet created a PR for it.
```
gem "kitchen-docker", :git => 'https://github.com/peterabbott/kitchen-docker.git', :branch => 'windows-integration' 
```

**Comment by dshah001** [comment]
Hi Peter,

Thanks for the update. I tried updating the gem, still get same error.

C:\Users\dshah001\kitchen\docker>bundle install
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and
installing your bundle as root will break this application for all non-root
users on this machine.
Using ffi 1.9.8
Using win32-process 0.7.5
Using win32-api 1.5.3
Using windows-api 0.4.4
Using windows-pr 1.2.4
Using mixlib-shellout 2.0.1
Using net-ssh 2.9.2
Using net-scp 1.2.1
Using safe_yaml 1.0.4
Using thor 0.19.1
Using test-kitchen 1.3.1
Using kitchen-docker 1.7.0 from https://github.com/peterabbott/kitchen-docker.git (at windows-integration)
Using bundler 1.6.2
Your bundle is complete!
Use `bundle show [gemname]` to see where a bundled gem is installed.

C:\Users\dshah001\kitchen\docker>kitchen list
       The system cannot find the path specified.
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::UserError
>>>>>> Message: You must first install the Docker CLI tool http://www.docker.io/gettingstarted/
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration

Does not seem to have it resolved yet..

**Comment by peterabbott** [comment]
It had been working for @markl17

Other things to try, make sure the docker executable is on your path. That either DOCKER_HOST is set or the socket driver attribute is set.

Failing that, try and set the binary property to 'docker.exe' (assuming is on path) or the absolute location.

For example:
```
driver:
  name: docker
  binary: docker.exe # <--because you are on windows
  socket: "tcp://<boot2docker ip>:<boot2docker port>" #<-- ideally use DOCKER_HOST to be portabe
  use_sudo: false
```

This comment has a bit of an example (although test suite formatting is a bit out, the driver config is right
https://github.com/peterabbott/kitchen-docker/issues/1#issuecomment-75583721

**Comment by dshah001** [comment]
Thanks Peter. Based on your example above, I updated my driver docker.rb to use > NUL 2>&1 which fixed the issue.

  def verify_dependencies
        run_command("#{config[:binary]} > NUL 2>&1", :quiet => true, :use_sudo => false)
        rescue

IC:\Users\dshah001\kitchen\docker>kitchen list
Instance             Driver  Provisioner  Last Action
default-ubuntu-1204  Docker  ChefSolo     <Not Created>
default-centos-64    Docker  ChefSolo     <Not Created>

C:\Users\dshah001\kitchen\docker>

**Comment by peterabbott** [comment]
I have also ported to a new gem that includes this fix

https://rubygems.org/gems/kitchen-docker_ssh

If you update the Gemfile with 

```
gem "kitchen-docker_ssh"
```

**Comment by looztra** [comment]
@peterabbott could you please be more specific about the kitchen-docker_ssh? Is it something completely different from kitchen-docker or is it just the same driver built from your "windows-integration" branch?

**Comment by peterabbott** [comment]
It is the same driver and includes a few fixes, including the windows
integration one.

There is also a fix to make the output a less chatty, the ability to use
with CircleCI and support for --cap-add and --cap-drop flags.

I'm open to add other PR from the driver. The idea was to have a driver
that was updated and published more often than every 6 months.

**Comment by looztra** [comment]
Thanks for the answer.

I have feedback, do we continue in this issue or in your fork? (https://github.com/peterabbott/kitchen-docker)

**Comment by peterabbott** [comment]
I guess move it to the fork or
https://github.com/peterabbott/kitchen-docker_ssh

**Comment by tknerr** [comment]
Hey @peterabbott  @looztra, how about merging the windows specific fixes back to kitchen-docker proper?

Docker 1.6.0 will ship with native docker CLI on windows. Probably more kitchenci users will run into this.

@peterabbott is this intended to be a permanent fork otherwise?

**Comment by peterabbott** [comment]
@tknerr PR in, best guest as to when it will get merged though.

**Comment by tknerr** [comment]
Thanks @peterabbott!

On a quick review I don't see the `> NUL` fix here in `verify_dependencies`:
https://github.com/peterabbott/kitchen-docker/blob/upstream/lib/kitchen/driver/docker.rb#L71

That was the one I was stuck at at first. Now I need to properly build docker.exe again so that it shows the version information first...

**Comment by peterabbott** [comment]
sorry, looks like my bad and was lazy and didnt check the code. 

Thought I had already merged the change into the upstream branch (branch I use to verify changes pulled down). I'll merge the proper changes in a little bit.

**Comment by markl17** [comment]
hi peter i have a very simple kitchen script it succeeds in making a machine on AWS but hangs on trying to get in it i tried ssh into it from another window with out any problems any ideas 
---
driver:
  name: ec2
  aws_access_key_id: 'something'
  aws_secret_access_key: 'something'
  aws_region: 'us-east-1'
  image_id: 'ami-d85e75b0'
  socket: "tcp://192.168.59.103:2376"
  binary: docker.exe
driver_config:
  require_chef_omnibus: 10.18.2
  
  run_command: /usr/sbin/sshd -D -o UseDns=no UsePam=no
  use_sudo: false
  provisioner_command:
    - chown kitchen.kitchen /tmp/kitchen
provisioner:
  name: chef_zero
  client_rb:
    no_proxy: 127.0.0.1
platforms:
  - name: dev
    driver_config:
      image: "repo:ubuntu"
      platform: "ubuntu"
suites:
  - name: webserver
    date_bags_path: databags
    roles_path: kitchen_roles
    run_list:
      - recipe[apache]
  - name: db01
    data_bags_path: databags
    roles_path: kitchen_roles

**Comment by peterabbott** [comment]
Not technically a thing with the docker driver but, could it be the
security permissions assigned to the group? Maybe port 22 is not open? I'd
check the security group config just to make sure the ssh key you expect is
linked to the box and port 22 is open. I remember I had some issues getting
it working with EC2, it was mostly that I was missing configuration
settings.

I'd generate a new kitchen config file from clean to make sure the docker
driver config settings are messing with the aws ones. In a clean directory
run `kitchen init` and update as required.

**Comment by markl17** [comment]
as I suspected it is trying to log in as root 
       Waiting for SSH service on ec2-52-7-90-9.compute-1.amazonaws.com:22, retr
ying in 3 seconds
       Waiting for SSH service on ec2-52-7-90-9.compute-1.amazonaws.com:22, retr
ying in 3 seconds
$$$$$$ [SSH] connection failed, terminating (#<Net::SSH::AuthenticationFailed: A
uthentication failed for user root@ec2-52-7-90-9.compute-1.amazonaws.com>)

**Comment by peterabbott** [comment]
There is a ssh user setting for the ec2 driver you can use to define. It
should be pulling it from the ec2 driver metadata but my experience is that
it is not always accurate.

Pete

**Comment by markl17** [comment]
i added transport: ubuntu to my script i dont think it changed anything

**Comment by jeffreycoe** [comment]
We recently released version 2.10 which includes Windows support. Please open a new issue if you encounter any problems with using this release.
